iceoryx2-services/discovery/src/service_discovery/service.rs

```diff
@@ -0,0 +1,350 @@
+                        iceoryx2::port::publisher::PublisherCreateError::UnableToCreateDataSegment => CreationError::PublisherCreationError,
+                    }})?
+            );
+        }
+
+        let mut notifier = None;
+        if discovery_config.send_notifications {
+            let event = node
+                .service_builder(service_name())
+                .event()
+                .max_listeners(discovery_config.max_listeners)
+                .max_notifiers(1)
+                .open_or_create()
+                .map_err(|_| CreationError::ServiceCreationFailure)?;
+
+            let port = event.notifier_builder().create().map_err(|e| match e {
+                iceoryx2::port::notifier::NotifierCreateError::ExceedsMaxSupportedNotifiers => {
+                    CreationError::NotifierAlreadyExists
+                }
+            })?;
+            notifier = Some(port);
+        }
+
+        let mut tracker = Tracker::<S>::new();
+
+        if discovery_config.sync_on_initialization {
+            tracker.sync(iceoryx_config);
+        }
+
+        Ok(Service::<S> {
+            discovery_config: discovery_config.clone(),
+            iceoryx_config: iceoryx_config.clone(),
+            _node: node,
+            publisher,
+            notifier,
+            tracker,
+        })
+    }
+
+    /// Processes service changes and emits events/notifications.
+    ///
+    /// This function should be called periodically to detect changes in the service
+    /// landscape and emit appropriate events and notifications. When services are
+    /// added or removed, the provided callback functions are invoked.
+    ///
+    /// # Parameters
+    ///
+    /// * `on_added` - Callback function that is called for each service that was added
+    /// * `on_removed` - Callback function that is called for each service that was removed
+    ///
+    /// # Returns
+    ///
+    /// A result containing `()` if successful.
+    ///
+    /// # Errors
+    ///
+    /// Returns a `SpinError` if there was an error publishing events or sending
+    /// notifications.
+    pub fn spin<
+        FAddedService: FnMut(&ServiceDetails<S>),
+        FRemovedService: FnMut(&ServiceDetails<S>),
+    >(
+        &mut self,
+        mut on_added: FAddedService,
+        mut on_removed: FRemovedService,
+    ) -> Result<(), SpinError> {
+        // Detect changes
+        let (added_ids, removed_services) = self.tracker.sync(&self.iceoryx_config);
+        let changes_detected = !added_ids.is_empty() || !removed_services.is_empty();
+
+        // Publish
+        for id in &added_ids {
+            if let Some(service) = self.tracker.get(id) {
+                if !self.discovery_config.include_internal
+                    && ServiceName::is_internal(service.static_details.name())
+                {
+                    continue;
+                }
+                self.publish(&Discovery::Added(service.static_details.clone()))
+                    .unwrap();
```

Comment:
I would not use `unwrap` here but just print an error if `publish` fails. Also, `publish is used two times and both of the times `unwrap` is called. Does `publish` need to return a `Result` if the caller just dismisses the result?